frame/result: allow differing TableSpecs in PreparedMetadata

wprzytula opened the pull request:
As noted in #1134, when preparing batches containing requests to multiple to different tables, the PreparedMetadata received upon preparation contains differing TableSpecs (i.e., TableSpecs with more than table mentioned).
This fails the check that we introduced with ResultMetadata in mind: we've been assuming that all TableSpecs are the same, because ScyllaDB/ Cassandra has no support for JOINs.
Not to fail preparation of cross-table batches, the check is now only turned on for ResultMetadata and turned off for PreparedMetadata.

Fixes: #1134

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

muzarski reviewed (commented):
Looks good, but can we add a regression test for the scenario mentioned in issue (i.e., where we prepare a batch statement with bind markers)?

wprzytula commented:
Rebased on main.

muzarski commented on scylla/src/statement/batch.rs:
```diff
@@ -178,6 +178,12 @@ impl From<&str> for BatchStatement {
     }
 }
 
+impl From<String> for BatchStatement {
+    fn from(s: String) -> Self {
+        BatchStatement::Query(Query::from(s))
+    }
+}
+
```
Was this change required to implement the test (it's in the same commit)?

wprzytula replied:
Good catch. I included this by accident, after I noticed such implementation is missing.

wprzytula replied:
I'm move this implementation out of this PR.

muzarski commented on scylla/src/transport/session_test.rs:
```diff
@@ -537,6 +537,42 @@ async fn test_batch() {
+#[tokio::test]
+async fn test_batch_to_multiple_tables() {
+    setup_tracing();
+    let session = Arc::new(create_new_session_builder().build().await.unwrap());
```
nit: is Arc required here?

wprzytula replied:
Not at all, indeed.

Lorak-mmk reviewed (changes requested):
Now that the metadata supports zero-copy deserialization, what is the benefit of having this check at all? If the only benefit is that some objects are a bit smaller, then this is imo not enough to justify additional complexity.
Diverging from CQL protocol have just hurt us, so we should avoid letting it hurt us again in the future.

wprzytula commented:
> Now that the metadata supports zero-copy deserialization, what is the benefit of having this check at all? If the only benefit is that some objects are a bit smaller, then this is imo not enough to justify additional complexity. Diverging from CQL protocol have just hurt us, so we should avoid letting it hurt us again in the future.

OK, I'll remove the check at all.

Lorak-mmk commented:
Ok. If there are other considerations here (e.g. simpler API) then it may be worth keeping the check. So the question remains - is it only a performance opt or something more?

wprzytula commented:
> Ok. If there are other considerations here (e.g. simpler API) then it may be worth keeping the check. So the question remains - is it only a performance opt or something more?

What is the "it" that you are referring to? What do you consider to be a performance optimisation? Because the check itself is for sure not an optimisation.

wprzytula commented:
v2:
- addressed @muzarski's comments,
- deleted the check (and the corresponding error variant) completely. All things considered, it's no use having it, especially once we established that we are not going to decouple TableSpec from ColumnSpec. We also no longer have strong motivation for it, because the ResultMetadata is now always deserialized in a borrowed form, which is much more lightweight.

A step that should actually be performed in that regard is pushing https://github.com/scylladb/scylladb/issues/17788 forwards, because the current way it's done is a dramatic waste.

Lorak-mmk commented:
> > Ok. If there are other considerations here (e.g. simpler API) then it may be worth keeping the check. So the question remains - is it only a performance opt or something more?
> 
> What is the "it" that you are referring to? What do you consider to be a performance optimisation? Because the check itself is for sure not an optimisation.

Sorry, I was writing the comment quickly and hastily.
The thing is I don't remember all the discussion and API changes around this.

What I meant is that if assuming the same table spec for columns gives us some API advantages, like being able to expose a table spec directly on query response (compared to expising it on the column metadata of the response), then it may be worth considering.

Getting rid of the assumption also lets us get rid of some types (like ColumnSpecView / ColumnSpecs), right? Then imo it is argument for that because it removes and simplifies code.

wprzytula commented:
What I like about `ColumnSpecs` is that it contains logic for `get_by_name` retrieval, which is absent on plain `&[ColumnSpec]`. I'd like to keep `ColumnSpecs` as such in our codebase. We could indeed remove `TableSpecView` and `ColumnSpecView`, and also provide an `inner` method for accessing `&[ColumnSpec]` from `ColumnSpecs`.

Lorak-mmk commented:
> What I like about `ColumnSpecs` is that it contains logic for `get_by_name` retrieval, which is absent on plain `&[ColumnSpec]`. I'd like to keep `ColumnSpecs` as such in our codebase. We could indeed remove `TableSpecView` and `ColumnSpecView`, and also provide an `inner` method for accessing `&[ColumnSpec]` from `ColumnSpecs`.
> 
> cc @muzarski

Ok, makes sense.

wprzytula commented:
This is API-breaking due to an error variant being removed. I suggest leaving the error variant for now under a TODO and removing it in the next major release. cc @Lorak-mmk

Lorak-mmk commented:
Yes, lets do that so we can release 0.15.1

wprzytula commented:
> This is API-breaking due to an error variant being removed. I suggest leaving the error variant for now under a TODO and removing it in the next major release. FYI @Lorak-mmk

Done.

muzarski commented:
Actually - one more nit: aren't docstrings/comment related to this function outdated now?

wprzytula commented:
> Actually - one more nit: aren't docstrings/comment related to this function outdated now?

Yees, good catch!

wprzytula commented:
> Actually - one more nit: aren't docstrings/comment related to this function outdated now?

Fixed.

muzarski commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -977,55 +977,24 @@ fn mk_col_spec_parse_error(
-            &mut known_table_spec,
-            col_idx,
-        )?;
+        let table_spec = match global_table_spec {
+            // If global table spec was provided, we simply clone it to each column spec.
+            Some(ref known_spec) if global_table_spec.is_some() => known_spec.clone(),
```
nit: we match against `global_table_spec` and have a match arm: `Some(_) if [...].is_some()`

muzarski replied:
I'm surprised clippy did not catch it
